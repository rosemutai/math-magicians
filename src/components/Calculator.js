import React from 'react';
import Display from './Display';
import Button from './Button';

const btnValues = [
  ['AC', '+/-', '%', '/'],
  [7, 8, 9, 'X'],
  [4, 5, 6, '-'],
  [1, 2, 3, '+'],
  [0, '.', '='],
];

const Calculator = () => (
  <div className="calc-homepage w-1/4 my-16 mx-auto bg-gray-200">
    <Display />
    <div className="grid grid-cols-4 grid-rows-5 gap-1">
      {btnValues.flat().map((item, index) => (
        <div key={item} className={`${index === 16 ? 'col-span-2' : 'col-span-1'}`}>
          <Button value={item} />
        </div>
      ))}
    </div>
  </div>
);

export default Calculator;
