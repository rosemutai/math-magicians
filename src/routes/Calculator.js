import { useState } from 'react';
import calculate from '../logic/calculate';

const greyButton = {
  style: 'bg-gray-100 py-3 px-6 text-black text-center font-regular',
};

const orangeButton = {
  style: 'bg-orange-400 py-3 px-6 text-black text-center font-regular',
};

const Calculator = () => {
  const [displayValue, setDisplayValue] = useState({ total: null, next: null, operation: null });

  const handleOnClick = (e) => {
    const valueClicked = e.target.textContent;
    const res = calculate(displayValue, valueClicked);
    setDisplayValue(res);
  };
  return (
    <div className="calc flex justify-between">
      <h1 className="text-3xl font-bold my-16 pl-12">Lets do some math</h1>
      <div className="calc-homepage w-1/4 my-16 mx-auto bg-gray-200">
        <div className="display-area px-2 py-4 flex justify-end bg-gray-400/75 w-full">
          <h2 className="display-value font-semibold text-white" data-testid="display-value">
            {displayValue.next || displayValue.total || 0}
          </h2>
        </div>
        <div className="grid grid-cols-4 grid-rows-5 gap-1">
          {/* row one */}
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>AC</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>+/-</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>%</button>
          <button type="button" className={`${orangeButton.style}`} onClick={handleOnClick}>รท</button>

          {/* row two */}
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>7</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>8</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>9</button>
          <button type="button" className={`${orangeButton.style}`} onClick={handleOnClick}>x</button>

          {/* row three */}
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>4</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>5</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>6</button>
          <button type="button" className={`${orangeButton.style}`} onClick={handleOnClick}>-</button>

          {/* row two */}
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>1</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>2</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>3</button>
          <button type="button" className={`${orangeButton.style}`} onClick={handleOnClick}>+</button>

          {/* row two */}
          <button type="button" className={`${greyButton.style} col-span-2`} onClick={handleOnClick}>0</button>
          <button type="button" className={`${greyButton.style}`} onClick={handleOnClick}>.</button>
          <button type="button" className={`${orangeButton.style}`} onClick={handleOnClick}>=</button>

        </div>
      </div>
    </div>
  );
};

export default Calculator;
