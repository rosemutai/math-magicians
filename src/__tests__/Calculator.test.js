import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';
import Calculator from '../routes/Calculator';

describe('Calculator', () => {
  it('renders Calculator component correctly', () => {
    const { container } = render(
      <Router>
        <Calculator />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });
});

describe('Calculator buttons work', () => {
  it('should update display value when buttons are clicked', () => {
    const { getByText, getByTestId } = render(
      <Router>
        <Calculator />
      </Router>,
    );

    const displayValue = getByTestId('display-value');

    fireEvent.click(getByText('5'));
    expect(displayValue.textContent).toBe('5');

    fireEvent.click(getByText('-'));
    expect(displayValue.textContent).toBe('5');

    fireEvent.click(getByText('2'));
    expect(displayValue.textContent).toBe('2');

    fireEvent.click(getByText('='));
    expect(displayValue.textContent).toBe('3');
  });
});
