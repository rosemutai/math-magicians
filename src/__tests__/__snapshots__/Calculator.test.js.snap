// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator renders Calculator component correctly 1`] = `
<div>
  <div
    class="calc flex justify-between"
  >
    <h1
      class="text-3xl font-bold my-16 pl-12"
    >
      Lets do some math
    </h1>
    <div
      class="calc-homepage w-1/4 my-16 mx-auto bg-gray-200"
    >
      <div
        class="display-area px-2 py-4 flex justify-end bg-gray-400/75 w-full"
      >
        <h2
          class="display-value font-semibold text-white"
          data-testid="display-value"
        >
          0
        </h2>
      </div>
      <div
        class="grid grid-cols-4 grid-rows-5 gap-1"
      >
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          AC
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          +/-
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          %
        </button>
        <button
          class="bg-orange-400 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          รท
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          7
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          8
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          9
        </button>
        <button
          class="bg-orange-400 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          x
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          4
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          5
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          6
        </button>
        <button
          class="bg-orange-400 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          -
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          1
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          2
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          3
        </button>
        <button
          class="bg-orange-400 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          +
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular col-span-2"
          type="button"
        >
          0
        </button>
        <button
          class="bg-gray-100 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          .
        </button>
        <button
          class="bg-orange-400 py-3 px-6 text-black text-center font-regular"
          type="button"
        >
          =
        </button>
      </div>
    </div>
  </div>
</div>
`;
